staging/src/k8s.io/dynamic-resource-allocation/cel/compile_experimental.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/ext"
+	resourceapi "k8s.io/api/resource/v1beta1"
+	"k8s.io/apimachinery/pkg/util/version"
+	apiservercel "k8s.io/apiserver/pkg/cel"
+	"k8s.io/apiserver/pkg/cel/environment"
+	"k8s.io/apiserver/pkg/cel/library"
+)
+
+const (
+	multiAllocVar = "allowMultipleAllocations"
+)
+
+var (
+	// A variant of BoolType with a known maximum size.
+	allowMultipleAllocationsType = withMaxElements(apiservercel.BoolType, 1)
+)
+
+var (
+	lazyExperimentalCompilerInit sync.Once
+	lazyExperimentalCompiler     *experimentalCompiler
+)
+
+func GetExperimentalCompiler() *experimentalCompiler {
+	lazyExperimentalCompilerInit.Do(func() {
+		lazyExperimentalCompiler = newExperimentalCompiler()
+	})
+	return lazyExperimentalCompiler
+}
+
+// Device defines the input values for a CEL selector expression.
+type ExperimentalDevice struct {
```

Comment:
I'm having trouble understanding why we're setting up an entire parallel "experimental" CEL universe ... shouldn't these be using normal mechanisms for becoming available to CEL expressions based on version / feature gate enablement?

the `VersionedOptions` IntroducedVersion and FeatureEnabled fields should allow you to introduce the new type declaration with the new fields conditionally if the version is new enough and the feature gate is enabled

Reply:
Do you mean that I can still keep the rest as previous except defining a new deviceType with multiAllocVar and add a new VersionedOption in the list with version >= 1.34, then use the new deviceType instead of present one?

Reply:
Discussed a bit with @sunya-ch, new version coming soon without the `compile_experimental.go` (not needed).

I am not sure whether the field should be feature gated. Suppose a CEL expression using the field gets stored. Now the feature gets turned off. If the field isn't in the environment anymore because of a feature gate check, compiling the stored expression fails.

I think we have to allow the field unconditionally, but then only evaluate it with devices which don't use consumable capacity. That's the if check in the allocator which filters out unsupported devices.

Reply:
Resolved this thread by changing to use VersionedOptions